redbpf-probes/src/maps.rs

```diff
@@ -392,6 +392,150 @@ impl<T> PerfMap<T> {
+    pub struct RingBufMap<T> {
+        def: bpf_map_def,
+        _event: PhantomData<T>,
+    }
+
+    impl<T> RingBufMap<T> {
+        /// Creates a ring buffer map with the specied maximum number of elements.
+        ///
+        /// `max_entries` must be a power of 2 value and must be page aligned.
+        pub const fn with_max_entries(max_entries: u32) -> Self {
+            Self {
+                def: bpf_map_def {
+                    type_: bpf_map_type_BPF_MAP_TYPE_RINGBUF,
+                    key_size: 0,
+                    value_size: 0,
+                    max_entries,
+                    map_flags: 0,
+                },
+                _event: PhantomData,
+            }
+        }
+
+        /// Copy data to ring buffer.
+        ///
+        /// The ring buffer can hold up to `max_entries` events, see `with_max_entries`.
+        /// No loss reporting occurs in user space. Monitoring must be done by the BPF
+        /// program.
+        ///
+        /// This method will wake up the polling thread if it is currently waiting on the
+        /// ring buffer. If this isn't desired, see `output_with_flags`.
+        #[inline]
+        pub fn output(&mut self, data: &T) -> i64 {
+            self.output_with_flags(data, RingBufMapFlags::default())
+        }
+
+        /// Copy data to ring buffer specifying wake options in the given
+        /// `RingBufMapFlags`.
+        #[inline]
+        pub fn output_with_flags(&mut self, data: &T, flags: RingBufMapFlags) -> i64 {
+            unsafe {
+                bpf_ringbuf_output(
+                    &mut self.def as *mut _ as *mut c_void,
+                    data as *const _ as *mut c_void,
+                    mem::size_of::<T>() as u64,
+                    flags.into(),
+                )
+            }
+        }
+
+        /// Reserve sample in ring buffer.
+        ///
+        /// This returns ref to raw pointer to reserved memory in the ring buffer. The
+        /// pointer should either be `submit`ted or `discard`ed, otherwise the BPF
+        /// program will fail to load with "Unreleased reference".
+        #[inline]
+        pub fn reserve(&mut self) -> Option<&mut T> {
+            unsafe {
+                (bpf_ringbuf_reserve(
+                    &mut self.def as *mut _ as *mut c_void,
+                    mem::size_of::<T>() as u64,
+                    0,
+                ) as *mut T)
+                    .as_mut()
+            }
+        }
+
+        /// Submit reserved sample.
+        #[inline]
+        pub fn submit(&mut self, data: &mut T) {
+            self.submit_with_flags(data, RingBufMapFlags::default())
+        }
+
+        /// Submit reserved sample specifying wake options in the given `RingBufMapFlags`.
+        #[inline]
+        pub fn submit_with_flags(&mut self, data: &mut T, flags: RingBufMapFlags) {
+            unsafe { bpf_ringbuf_submit(data as *mut _ as *mut c_void, flags.into()) }
+        }
```

Comment:
This auxilliary struct could also help contain the received pointer. I think we could expose the ptr -> ref conversion through an `unsafe fn`, or through a `Result<&mut T, E>` gate, where the accessor function validates pointer alignment before creating the reference.

Reply:
These are great suggestions. Using auxiliary struct with the Drop trait instead of an explicit discard should make probe logic cleaner and simpler.

Reply:
Still outstanding. How do I go about validating pointer alignment here?

Reply:
How about passing value and write it to the dest using `ptr::write_unaligned`?

Reply:
Doesn't this defeat the optimizations of reserve/submit? The idea is to avoid creating temporary structures and avoiding copies altogether.

Can we make this the responsibility of the user of the API instead?

I created the auxiliary type as discussed. The user can chose to use the unsafe `as_mut` helper, or access the ptr directly from the struct. What do you think?

Reply:
For reference, this is how I'm using the API now. I think it looks better now!

```
+        let mut payload = NEW_TASKS.reserve().ok_or(ProbeError::RingBufReserveError)?;
+        let event = payload.as_mut();
+        event.pid = task.0.pid as u64;
+        event.ppid = bpf_get_current_pid_tgid() >> 32;
+        payload.submit();
```